SDK Trace: Adding span collection limits

toumorokoshi opened the pull request:
To help safeguard against a common error, adding default limits
to the number of events per span in the SDK.

Fixes #182

## Changes

add default limits to the number of events per span in the SDK

- [x] update changelog
- [x] update spec compliance matrix

iNikem commented on specification/trace/sdk.md:
```diff
@@ -236,7 +236,27 @@ Thus, the SDK specification defines sets of possible requirements for
+events that would increase the size of each collection beyond a limit configured
+in the `TracerProvider`, or for the span specifically.
+
+There MUST be a log message emitted by the SDK for each object discarded.
```
I am uneasy about this requirement. First, AFAIK there is not any agreement right now about what and how SDK should log. Second, if we try to protect against unexpected huge number of events added to span, then we should not write unexpectedly huge number of log records.

toumorokoshi replied:
Sounds good, but I think it would be good to require *some* information to help someone understand any missing events.

How about requiring that a message is logged when a collection discards for the first time? That will result in at most 3 logs per erroneous span (assuming each collection is filled).

Oberon00 replied:
I don't think we can require even that, as it would basically forbid the SDK to allow turning off logging. I think we should say, that a it SHOULD be possible to *configure* a message to be logged.

bogdandrutu replied:
I am proposing to just keep track of the number of different things that we drop and push that to the exporters, then every exporter can decide what to do.

carlosalberto replied:
>  that a it SHOULD be possible to configure a message to be logged

+1

carlosalberto replied:
> I am proposing to just keep track of the number of different things that we drop and push that to the exporters, then every exporter can decide what to do.

Also a good alternative, either way.

toumorokoshi replied:
> as it would basically forbid the SDK to allow turning off logging.

Is it not generally implicit that any logs emitted by opentelemetry could be disabled? If not, I wonder how one would appropriately express "you must emit the logs, but it can disabled". I'm not sure "SHOULD" expresses that intent well.

> I am proposing to just keep track of the number of different things that we drop and push that to the exporters, then every exporter can decide what to do.

I think tracking and emitting is a good idea. Would exporters be the right place to add that complexity? I could imagine, ideally, that these should be metrics emitted. But I'm not sure if we want the complexity that comes with span + metrics working in concert like that.

iNikem replied:
When reading Bogdan's proposal, I though first about metrics as well. Not sure if this PR is the right place to start introducing this, but I think we currently underuse internal metrics.

toumorokoshi replied:
how about a separate issue to discuss the exporting of metrics from trace SDK components? I don't see a precedent in the SDK, and I worry that may divert this PR into a much bigger topic.

toumorokoshi replied:
started #959, and I'll update the wording here as @Oberon00 suggested.

toumorokoshi replied:
wording updated

iNikem commented on specification/trace/sdk.md:
```diff
@@ -236,7 +236,27 @@ Thus, the SDK specification defines sets of possible requirements for
+The following interfaces MUST exist:
+
+- `TracerProvider`: a method to set a maximum collection size
+- `TracerProvider`: a method to retrieve the maximum collection size
```
This requires only the ability to set a common limit for everything: attributes, events and links, correct? Should we mention that SDK can provide individual configuration options for each type?

toumorokoshi replied:
Agreed that having individual limits per collection is valuable, I was considering that as an addition in a future spec version based on user feedback, but having an optional addition makes perfect sense.

bogdandrutu replied:
This should be on the TracerProvider implementation, not on the API correct?

toumorokoshi replied:
yes, correct. i'll take a look through and see if I can better clarify that.

toumorokoshi replied:
What's the right way to qualify that in this document? References to TracerProvider (API or SDK) are intermingled, even in this page:

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#tracer-creation

Should I change these references specifically to "SDK TracerProvider"?

Oberon00 replied:
Since this is the SDK document, I think just saying TracerProvider (as-is) should be fine.

toumorokoshi replied:
I'll close this as there hasn't been a better suggestion yet, please recommend an approach and I'm happy to address.

iNikem commented on specification/trace/sdk.md:
```diff
@@ -236,7 +236,27 @@ Thus, the SDK specification defines sets of possible requirements for
+- `TracerProvider`: a method to retrieve the maximum collection size
+- `Span`: a method to set a maximum collection size for the span
+
+The default for this limit SHOULD be 1000. A value of 0 indicates no limit.
```
When I tried to implement in Java SDK a similar limit for the _size_ of attribute values, there were a lot of arguments that 0 is in fact a bad "no limit" value. And that "-1" is a better one. 0 is potentially a valid size of collection, -1 is obviously not.

toumorokoshi replied:
Agreed, I debated that as 0 is still a common unlimited value in other systems. I'll update to -1.

toumorokoshi replied:
updated.

bogdandrutu commented on specification/trace/sdk.md:
```diff
@@ -236,7 +236,31 @@ Thus, the SDK specification defines sets of possible requirements for
+
+- `TracerProvider`: a method to set a maximum collection size
+- `TracerProvider`: a method to retrieve the maximum collection size
+- `Span`: a method to set a maximum collection size for the span
```
-1 on this. The `Span` is an API interface and this implementation detail should not leak into the API. I suggest to remove this for the moment.

toumorokoshi replied:
ok, sounds good. I was trying to find a way for users to override spans that they *know* will go above the limit.

It seems to me if someone is planning on adding an exception, they may want to do so per span, rather than up the limit for all spans.

toumorokoshi replied:
removed.

toumorokoshi commented on specification/trace/sdk.md:
```diff
@@ -236,7 +236,31 @@ Thus, the SDK specification defines sets of possible requirements for
+- `TracerProvider`: a method to retrieve the maximum collection size
+- `Span`: a method to set a maximum collection size for the span
+
+Implementations MAY additionally have configuration for limits on each
+collection, which would take precedence over the general collection limit.
```
each == attributes, events, and links. I'll update the wording to call them out specifically.

toumorokoshi replied:
replaced with more specific wording.

SergeyKanzhelev commented:
@tigrannajaryan @bogdandrutu @Oberon00 are you satisfied with the PR update? Can you pls re-review?

tigrannajaryan commented on specification/trace/sdk.md:
```diff
@@ -236,7 +237,34 @@ Thus, the SDK specification defines sets of possible requirements for
+events that would increase the size of each collection beyond a limit configured
+in the `TracerProvider`, or for the span specifically.
+
+There SHOULD be a log message emitted by the SDK for the first time a span has
+discarded a span due to a limit. This message should be emitted once for
```
```suggestion
There SHOULD be a log message emitted by the SDK for the first time a span is
discarded due to a limit. This message should be emitted once for
```

tigrannajaryan commented on specification/trace/sdk.md:
```diff
@@ -236,7 +237,34 @@ Thus, the SDK specification defines sets of possible requirements for
+
+There SHOULD be a log message emitted by the SDK for the first time a span has
+discarded a span due to a limit. This message should be emitted once for
+attributes, events, and links, and MUST be configurable on or off, with
```
>MUST be configurable on or off

Is configurability necessary?

toumorokoshi replied:
See https://github.com/open-telemetry/opentelemetry-specification/pull/942#discussion_r487719748 for the discussion on that. 

I believe many languages have configurable loggers so it doesn't seem like a difficult / unreasonable requirement, but please share details if you feel differently.

bogdandrutu reviewed (commented):
After reading this, I am not sure if you ask for a "bytes limit" or "number of attributes", "number of events", etc limits.

I would say that the first one may sound very interesting but it comes with some performance overhead and hard to implement it right. So I am more supportive of the "max number of attributes", etc.

Independent of what is suggested here, I think the fact that is not clear has to be fixed.

toumorokoshi commented:
> Independent of what is suggested here, I think the fact that is not clear has to be fixed.

Good point! I've fixed to element count.

I agree that bytes has some interesting ramifications and addressed the immediate concern of stressing memory, but maybe a follow-up in another PR.

reyang commented on specification/trace/sdk.md:
```diff
@@ -236,7 +237,34 @@ Thus, the SDK specification defines sets of possible requirements for
+- `TracerProvider`: a method to set a maximum collection count
+- `TracerProvider`: a method to retrieve the maximum collection count
+
+The default for this limit SHOULD be 1000 elements. A value of -1 indicates no
```
Why do we want to have `-1` (no limit) and what is the expected result when running out of resource?

iNikem replied:
E.g. if we have streaming span processor/exporter, which streams data as soon as possible, then unlimited makes sense. Implementing unlimited as `Long.MAX_VALUE` is clumsy.

toumorokoshi replied:
right, I think that's the alternative. Certainly we could not honor a -1 / unlimited sentinel value, But someone will just choose some insanely large number anyway.

I would argue providing some semantic value to indicate the choice would be better than an implicit value, but I'm not strongly opinionate.

toumorokoshi replied:
The PR is much looser now, so I think guidelines around limits can be debated when there is a spec "MUST", rather than a "MAY".

reyang commented on specification/trace/sdk.md:
```diff
@@ -236,7 +237,34 @@ Thus, the SDK specification defines sets of possible requirements for
+The following interfaces MUST exist:
+
+- `TracerProvider`: a method to set a maximum collection count
+- `TracerProvider`: a method to retrieve the maximum collection count
```
I'm worried about going down this path.
In practice everything has a limit (e.g. the maximum length of a span name, the maximum number of span attributes), we probably don't want to end up having everything configurable.

toumorokoshi replied:
I can see that perspective. I guess I would worry about enabling cases that need more than the default limit. 

We could also drop the configurability for now, and let someone file an issue with a use case if they want to see the value change.

Does anyone else want to share their thoughts?

iNikem replied:
I think the configurability suggested in this PR is useful and should remain. Use-case is right in the description: to prevent uncontrolled memory usage.

SergeyKanzhelev replied:
Like with other parameters @reyang listed, some default (or dynamic default) may address the uncontrolled memory growth scenarios. The question is what's the use case of making it configurable. With the high enough defaults, limiting number of arguments to a smaller number can be done in exporter, unless I'm missing something. The actual question is what's the scenario where limit needs to be removed. @iNikem can you please explain how do you envision this scenario: https://github.com/open-telemetry/opentelemetry-specification/pull/942#discussion_r490001964 and why limit should be removed then?

iNikem replied:
@SergeyKanzhelev I am afraid I don't understand your question. During various discussion I have heard that we should support (or at least not prevent) streaming span processing pipelines: streaming span processor and exporter. In streaming case one don't want and probably cannot enforce max size constraint on the number of span attributes or events. Thus it make sense to have an option to configure unlimited sizes.

What do you mean by "dynamic default"?

I don't see any downside in allowing an end-user to configure data collection to their liking. Especially in case like this, where we want to limit something. We have to give end user an opportunity to change those limits.

SergeyKanzhelev replied:
So there are different ways to go about it. In OpenCensus the defaults are quite low. So it is only natural to provide a way to bump it a little bit. In OpenTelemetry making defaults low seems to be against the idea of interoperability. So my assumption was that the default would be quite high to allow exporters to decide on cutting extra events or attributes the way it prefers. There might even be dynamic defaults which shrinks as memory pressure increases. So visibly unlimited number.

Now if the defaults are quite high, the only reason to declare configurability in a spec seems to be to allow unlimited number. With the unlimited number, I wonder what kind of streaming scenario is envisioned here. I thought it's something about GRPC channels or SinglaR. 

If this is streaming SDK, than it is completely different SDK that is not spec'd at all. The only statement OTel is making is that OTel API would allow to implement streaming SDK. In case of this SDK, making that limit configurable doesn't seem logical. Just keep it infinite shouldn't be a problem.

iNikem replied:
Do I understand you correctly, that you propose to have hard-coded limit of 1000 attributes/events/links per span without any means to configure that? This goes against all my experience with software that arbitrary end-users will use in unknown scenarios.

SergeyKanzhelev replied:
hardcoded limits are virtually everywhere. How many tracestate entries can you add? How many http headers, cookies =). Like what @reyang said above.

With the limit of `1000` it's almost like unlimited for many scenarios. This is why I wonder what do you envision as a scenario where lower or higher limits are needed? I believe that scenarios need to drive the decision on what's the requirement for every language and ultimately described in the spec.

toumorokoshi replied:
I personally side with having a configurable limit, but there's good arguments that not much else about OpenTelemetry is configurable today. 

Also the biggest value this PR can provide is to set a limit and get it in before GA, to set some limitation and protect against a vast majority of erroneous cases.

I'll file a follow-up issue, but reduce this change to a non-configurable limit of 1k.

toumorokoshi replied:
@SergeyKanzhelev @iNikem @reyang

iNikem replied:
I object to the current version with hard-coded limit of 1000 and no way to change that. Current version effectively _forbids_ SDKs to provide such configuration option, even if e.g. Java already provides that.

I can agree with the version which allows SDKs to provide an optional configuration for these limits.

toumorokoshi replied:
I think that's fair. I think whether the SDK forbids it is up to interpretation (based on your assumption upon omission of the topic), but agreed clarification can't hurt.

toumorokoshi replied:
added a *may*

andrewhsu commented:
from the issue triage meeting today with TC, if this PR is desired before GA, need 2 approvals an no disapprovals to merge, else this will be closed by end of day today @open-telemetry/technical-committee @open-telemetry/specs-approvers  @SergeyKanzhelev

jmacd reviewed (approved):
The reason I might object to this spec change is simply that it's a lot of work to require, but it's work that was inherited from the OpenCensus libraries in many cases, e.g., I believe these limits are already implemented for OTel-Go.

andrewhsu commented:
from the issue triage mtg today with TC, marking as `after-ga` because corresponding issue is also `after-ga`

iNikem commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,22 @@ Thus, the SDK specification defines sets of possible requirements for
+
+To protect against such errors, SDK Spans MUST discard attributes, links, and
+events that would increase the number of elements of each collection beyond
+1000.
```
```suggestion
1000. SDKs MAY provide a way to change this limit.
```

SergeyKanzhelev reviewed (approved):
I agree with the spirit of this change. .NET may not implement it for all collection by GA, but a single language should not block the spec. There are few comments on wording as well

SergeyKanzhelev commented on spec-compliance-matrix.md:
```diff
@@ -40,6 +40,7 @@ status of the feature is not known.
 |IsRecording                                   | + | +  | + | +    | +  | +    | + |    | + | +  |
 |Set status                                    | + | +  | + | +    | +  | +    | + | +  | + | +  |
 |Safe for concurrent calls                     | + | +  | + |      | +  | +    | + | +  | + | +  |
+|Discard events greater than collection limit  |   |    |   |      |    |      |   |    |   |    |
```
It's more than just events. Either use separate lines for different collections or mentione them all here. 

```suggestion
|Safe handling of unbounded growth of events, attributes and links collections  |   |    |   |      |    |      |   |    |   |    |
```

SergeyKanzhelev commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,22 @@ Thus, the SDK specification defines sets of possible requirements for
+
+To protect against such errors, SDK Spans MUST discard attributes, links, and
+events that would increase the number of elements of each collection beyond
+1000. SDKs MAY provide a way to change this limit.
```
```suggestion
the recommended limit of 1000 elements. SDKs MAY provide a way to change this limit.
```

toumorokoshi commented:
@SergeyKanzhelev do you think this could be merged? (pinging you as owner). I believe we have sufficient approvals, despite this not being a GA priority.

arminru commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,22 @@ Thus, the SDK specification defines sets of possible requirements for
+the recommended limit of 1000 elements. SDKs MAY provide a way to change this limit.
+
+There SHOULD be a log message emitted by the SDK for the first time a span is
+discarded due to a limit. This message should be emitted once for
+attributes, events, and links, and MUST be configurable on or off, with
```
Please clarify if this is meant to be once per process lifetime or once for each span. I suppose once per process should be sufficient as logging on each "overloaded" span could result in log spam.

toumorokoshi replied:
I was actually considering one per span (to help provide signal on specific spans which have lost values).

That said one per process may be enough.

tigrannajaryan replied:
I don't think "once per span" is good enough. We want to avoid flooding the logs. "once per process" can work. Event better, we can avoid being too prescriptive and instead just recommend log rate limiting. Some log libraries have rate limiting built-in, we don't need to invent it.

toumorokoshi replied:
resolving this, and starting another thread with an updated PR. 

I've gone your latest suggestion and just stating that logging should exist, but not be excessive.

arminru commented:
@open-telemetry/technical-committee should we add such a sanity limit to prevent app crashes due to instrumentation problems for the GA version?
This would be a breaking change if one relies on adding more items than the default limit and would at least require them to change the configuration (if even supported by the SDK) in that case.
Some SIGs already implemented such limits (e.g. [Java](https://github.com/open-telemetry/opentelemetry-java/blob/64c2b8cc65a023bdf6f1e09bfc73bbe20168b627/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java)), but I'm not sure if all do.

tigrannajaryan commented:
> @open-telemetry/technical-committee should we add such a sanity limit to prevent app crashes due to instrumentation problems for the GA version?

I do not quite like the hard stance taken by this proposal. I think the spec is overstepping into what arguably can be also seen as an implementation decision.

I also do not like that we prescribe only one way to limit resource exhaustion by telemetry data. Limiting the total number of elements is not the only possible way and arguably is not necessarily completely effective either because we do not simultaneously limit the size of the element.

Some SDK's may have alternate solutions to this problem, e.g. implement limits for the _total_ byte size rather then for the size of an individual element. By prescribing the specific behavior in the spec we prevent innovation in this area.

I would be more comfortable with this proposal if it was a _recommendation_.

toumorokoshi commented:
Thanks for the feedback! It's great to have more input.

> I also do not like that we prescribe only one way to limit resource exhaustion by telemetry data. 

Do you believe this will restrict people from adding other ways? I thought the spec was always a dictation of what is a requirement for an implementation, with anything not explicitly stated up to the liberty of the implementations.

> Some SDK's may have alternate solutions to this problem, e.g. implement limits for the total byte size rather then for the size of an individual element. By prescribing the specific behavior in the spec we prevent innovation in this area.

I think this is the tradeoff to be made: the con is forcing choices that are not appropriate for the implementation. the pro is standard behavior that's enables knowledge that transfers across language boundaries. We dictate several implementation details that could be argued to prevent innovation, such as dictating that the [behavior of the default span processor, and it's defaults](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#built-in-span-processors).

If the consensus (and I'd like to hear a few more chime in too) is "we don't know if this is the right behavior to dictate across implementations", I think it's fine to leave this PR be (or close). 

But I would argue that this is a practical implementation detail that would prevent a significant category of human errors. And the cost of differing implementations is end user confusion.

SergeyKanzhelev commented:
> Some SDK's may have alternate solutions to this problem, e.g. implement limits for the _total_ byte size rather then for the size of an individual element. By prescribing the specific behavior in the spec we prevent innovation in this area.

the proposal in current form is very toothless. `1000` is non-limit in real life. Just a statement that SDKs needs to watch out for these collections.

If anybody disagree with this PR - please mark it with "changes requested". Otherwise I will merge mid day (pacific time) tomorrow.

tigrannajaryan commented:
> Thanks for the feedback! It's great to have more input.
> 
> > I also do not like that we prescribe only one way to limit resource exhaustion by telemetry data.
> 
> Do you believe this will restrict people from adding other ways? I thought the spec was always a dictation of what is a requirement for an implementation, with anything not explicitly stated up to the liberty of the implementations.

@toumorokoshi Not restrict, but possibly discourage. There is limited amount of time SDK authors are willing to spend on this topic. As a SDK author if I spend effort to implement a spec requirement I am less likely to go ahead and implement another additional way to limit resource exhaustion. Developer's time is a finite resource too.

> > Some SDK's may have alternate solutions to this problem, e.g. implement limits for the total byte size rather then for the size of an individual element. By prescribing the specific behavior in the spec we prevent innovation in this area.
> 
> I think this is the tradeoff to be made: the con is forcing choices that are not appropriate for the implementation. the pro is standard behavior that's enables knowledge that transfers across language boundaries. We dictate several implementation details that could be argued to prevent innovation, such as dictating that the [behavior of the default span processor, and it's defaults](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#built-in-span-processors).

I agree that it is beneficial to have a good standard behavior. I am not convinced the proposed standard behavior is valuable enough to be enforced across all SDKs.

> If the consensus (and I'd like to hear a few more chime in too) is "we don't know if this is the right behavior to dictate across implementations", I think it's fine to leave this PR be (or close).
> 
> But I would argue that this is a practical implementation detail that would prevent a significant category of human errors. And the cost of differing implementations is end user confusion.

I would also like to hear more opinions. I am willing to be convinced.

tigrannajaryan reviewed (changes requested):
For reasons I posted in my previous comments I am putting a soft block on this to prevent premature merging, but I am open to be convinced.

iNikem commented:
I would like to point to yet another aspect of the current discussion. Even if this PR is not merged, it will be allowed for SDK to implement these or similar limits. Then it will be beneficial if different languages at least used the same configuration options for these limits. E.g. the same environment variable name. Is the specification open to augment `sdk-environment-variables.md` with configuration for such optional behaviour?

tigrannajaryan commented:
>Is the specification open to augment sdk-environment-variables.md with configuration for such optional behaviour?

Yes, I think so. I would prefer configurable behavior.

SergeyKanzhelev commented:
> > Is the specification open to augment sdk-environment-variables.md with configuration for such optional behaviour?
> 
> Yes, I think so. I would prefer configurable behavior.

Configurability was removed from this PR as we cannot formalize use cases. For instance, if somebody set the limit to 2 attributes, should it be supported? Why? If somebody sets the limit to infinity, should it be supported? No exporter will probably need those so why to allow to set it?

Configurability was important in OpenCensus as default was quite low. Like 10 or 20 as far as I remember. In OpenTelemetry I believe the general agreement that we need to start with much higher default. Which defeats the purpose of having this limits configurable on SDK side.

I see this PR in current form as a warning that something need to be done to control for unbound collections. Alternative to this PR is doing nothing - SDKs needs to be safe in general. Switching to controllable behavior would need to start with the scenarios.

tigrannajaryan commented:
>If somebody sets the limit to infinity, should it be supported? No exporter will probably need those so why to allow to set it?

There is a use case where collection of all telemetry is important: audit. In extreme cases it may be preferable to crash the actual operation if the telemetry that describes the operation cannot be exported. I understand that we do not necessarily care about this particular use case. 

>I see this PR in current form as a warning that something need to be done to control for unbound collections.

Makes sense. If it's a warning perhaps we can it a recommendation rather than hard requirement?

toumorokoshi commented:
I think there's a good comprise with the optional aspect of it. Aggregating the proposal and calling out how this is addressing concerns, how about:

- make it optional
    - addresses @SergeyKanzhelev's point about having something in the spec to bring awareness to this issue.
    - enables experimentation as per @tigrannajaryan's point.
- keep the configurability, as per @iNikem and @tigrannajaryan's ask. 
   - add suggested environment variables in sdk-environment-variables.md

I'm still concerned about this not being required, but I'd much prefer some awareness of the issue in the spec vs none at all. And it can be a good building block if we want to expand that into something more standardized.

iNikem commented:
> SDKs needs to be safe in general

I am sorry, what does it mean? SDKs cannot be safe by magic, they have to build themselves in such a way. This PR is one way to help/advice SDKs how to do that. SDKs have to implement some safety measures, this PR tries to make them similar across different languages.

Do we want to have ever language different? One has low limit, one has no limit, everyone using different configuration variables?

toumorokoshi commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,24 @@ Thus, the SDK specification defines sets of possible requirements for
+If there is a configurable limit, the SDK MAY honor the environment variables
+specified in [SDK environment variables](../sdk-environment-variables.md#span-collection-limits).
+
+There SHOULD be a log emitted to indicate to the user that an attribute, event,
```
@tigrannajaryan PTAL and see if this addresses your concerns around span collection limit logging.

tigrannajaryan commented:
Thank you @toumorokoshi

tigrannajaryan commented:
@SergeyKanzhelev @iNikem I re-requested your review since the PR is substantially different now.

toumorokoshi commented on spec-compliance-matrix.md:
```diff
@@ -40,9 +40,6 @@ status of the feature is not known.
 |Safe for concurrent calls                     | + | +  | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1157)    | +  | +    | + | +  | + | +  |
-|events collection size limit                  |   |    |   |      |    |      |   |    |   |    |
-|attribute collection size limit               |   |    |   |      |    |      |   |    |   |    |
-|links collection size limit                   |   |    |   |      |    |      |   |    |   |    |
```
re-added!

iNikem commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,24 @@ Thus, the SDK specification defines sets of possible requirements for
+
+To protect against such errors, SDK Spans MAY discard attributes, links, and
+events that would increase the number of elements of each collection beyond
+the recommended limit of 1000 elements. SDKs MAY provide a way to change this limit,
```
```suggestion
the recommended limit of 1000 elements. SDKs MAY provide a way to change this limit.
```

iNikem commented on specification/trace/sdk.md:
```diff
@@ -241,7 +242,24 @@ Thus, the SDK specification defines sets of possible requirements for
+events that would increase the number of elements of each collection beyond
+the recommended limit of 1000 elements. SDKs MAY provide a way to change this limit,
+
+If there is a configurable limit, the SDK MAY honor the environment variables
```
```suggestion
If there is a configurable limit, the SDK SHOULD honor the environment variables
```

If SDK decides to implement limits, there is no reason not to use common configuration options, is there?

toumorokoshi replied:
agreed, just debating how to signal the recommendation strength based on an already optional addition. But understood that we should use "SHOULD / MUST" with an optional clause.

arminru reviewed (approved):
LGTM once @iNikem's final suggestions are applied.

Thanks for carrying this through, @toumorokoshi!

toumorokoshi commented:
@arminru added!

toumorokoshi commented:
@SergeyKanzhelev could you do a final review, and merge if you feel it's appropriate? I believe I've addressed all outstanding comments.
